### ISSUE TITLE
ipv6 support

### ISSUE THREAD

**Issue by 0-kaladin** [Free-Software-for-Android/AdAway#667]
@dschuermann I see in RegexUtils.java there is ipv6 support.  But when I point adaway to this hosts file http://someonewhocares.org/hosts/ipv6/hosts that has domains in both ipv4 and ipv6 format like:
127.0.0.1 metrics.sun.com
::1 metrics.sun.com

I only see ipv4 making it into the parsed list.  I'm trying to follow the regex's but they are a tad complicated, do you know if this ever worked or just implemented on the backend?  I only ask because I wonder if ads will start being served from ipv6 and I can also ensure the local webserver will bind to localhost for ipv6 as well like ::1:80 but we'd need the ::1 localhost in the hosts file as well I think.

**Comment by dschuermann** [comment]
I think the ipv6 regex is only used for checking validity inside AdAway's white, black and redirection lists. Not when parsing the downloaded hosts file. Somewhere it is checked if a line starts with 127.0.0.1, thsi needs to be extended to also allow "::1", but nothing more.

**Comment by 0-kaladin** [comment]
hmmm, yea around line 100 of HostsParser.java I added a check for IPv6
```
if (currentIp.equals(Constants.LOCALHOST_IPv4)
                        || currentIp.equals(Constants.LOCALHOST_IPv6)
                        || currentIp.equals(Constants.BOGUS_IPv4)) {
                    mBlacklist.add(currentHostname);
                } else if (currentIp.equals(Constants.WHITELIST_ENTRY)) {
                    mWhitelist.add(currentHostname);
                } else if (mParseRedirections) {
                    mRedirectionList.put(currentHostname, currentIp);
                }
```

but it is still just adding 127.0.0.1 entries.  I bet the hostsparser isn't allowing duplicate hosts entries then.  Since with ipv6 support we would need
127.0.0.1 badhost.com
::1 badhost.com

**Comment by dschuermann** [comment]
Yup no duplicate entries are accepted. When writing the hosts file you could write a second line with the same hostname.

**Comment by 0-kaladin** [comment]
could but then doubling size of hosts file.  I guess ideally the Black list takes currentIp like Redirection List does so if your source file defined a 127.0.0.1 host.com and a ::1 other.com adaway would add it.  Currently every blacklist item gets added as 127.0.0.1 whether the source file has that, 0.0.0.0 or now even ::1.

Also (since i've been looking at the webserver) I don't see the benefit in having it listen on 443.  it can never return 200 ok as the security never gets setup.  so whether local webserver is running or not any https requests just get invalid ssl session response.  I'd leave with 443 for now but maybe remove if i can ever implement blacklist supporting ipv6 entries so i can have the webserver listen on [::1]:80 which i've tested:
with a ::1 www.google.com entry this returns great:
./curl -I -g -6 www.google.com
HTTP/1.1 200 OK
Content-Length: 0
Content-Type: text/plain

**Comment by dschuermann** [comment]
Hmm, answering on 443 with a wrong certificate or no certificate could be better than not answering at all, I am not sure what the user sees when we don't answer on 443. Is it loading endlessly?

Regarding https://github.com/Free-Software-for-Android/AdAway/commit/72d34b7910509c58b6ba6c46cee930caa6bad79d
Why have you removed the oom stuff?

**Comment by dschuermann** [comment]
Regarding the blacklisting of IPv6:
http://someonewhocares.org/hosts/ipv6/hosts simply includes ::1 for all hostnames as far as I see. So they don't even check if a host has IPv6 or not.
Thus, I don't see a reason to use the actual IP from the list instead of always using 127.0.0.1 plus a second line for ::1.
I would propose adding a preference to enable/disable IPv6 entries.

**Comment by 0-kaladin** [comment]
1.  I don't intentionally remove oom, i just start with a blank file when re-doing the webserver and go through a few iterations and tests and always forget to add it in.  I've run it for 2.5 days now and no oom kills the process, but I am not using memory starved devices :) .
I've added in back in case someone does have a 512 ram device on ICS or higher. 
https://github.com/Free-Software-for-Android/AdAway/commit/2dff8eb34fcdbcdd76aaec58f4112e12726f25bf
one modification is that I switched from oom_adj which was deprecated in some 2.6.x kernel version in favor of the oom_score_adj now recommended to use.  since android has been on 3.x kernel's since ICS and our API level is higher than that this is a good change for future proofing.

2.  Regarding 443, first the reason i ask is if I have local webserver listen on ::1 and 127.0.0.1 with port 80 only it uses in resident about 392k of ram.  When i add another port, 443, even if only for 127.0.0.1 the resident memory doubles to about 692k.  Which makes sense since with a different port you really need another instance of the server whereas binding to different interfaces with same port is able to reuse the same.  So that is the basis for wondering the value of 443.  
Moving on to my tests, no one gets loading endlessly so we are good there as that would be bad.

If I analyze the purpose of the webserver, it is to ensure you get a clean, 200 OK response, with no content so that the ads that may have loaded are now fully blank.  In many cases this even isn't helpful since many apps will not surface to the user level any errors from loading their ads.  But some do and there are different situations where it is preferable to get a response.  So using curl to get the headers of a request here is a port 80 request to a server in the hosts file without the webserver:
'''
./curl -I -v http://gh.com/
*   Trying 127.0.0.1...
* connect to 127.0.0.1 port 80 failed: Connection refused
* Failed to connect to gh.com port 80: Connection refused
* Closing connection 0
curl: (7) Failed to connect to gh.com port 80: Connection refused
'''
Turning on the 80 webserver it looks like this:
'''
./curl -I -v http://gh.com/
*   Trying 127.0.0.1...
* Connected to gh.com (127.0.0.1) port 80 (#0)
> HEAD / HTTP/1.1
> User-Agent: curl/7.40.0
> Host: gh.com
> Accept: */*
>
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Content-Length: 0
Content-Length: 0
< Content-Type: text/plain
Content-Type: text/plain
'''

Now lets look at HTTPS.  with 443, no webserver, response is the same as 80.  First connect line, you get connection refused (since nothing is on 127.0.0.1 to answer).
with the local webserver on 443 we get:
'''
./curl -I -k -v https://gh.com/
*   Trying 127.0.0.1...
* Connected to gh.com (127.0.0.1) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /data/local/ssl/certs
* TLSv1.0, TLS handshake, Client hello (1):
* Unknown SSL protocol error in connection to gh.com:443
* Closing connection 0
curl: (35) Unknown SSL protocol error in connection to gh.com:443
'''

so you get an initial connection since the webserver is there, but you then get an ssl error and then close connection.  So the question comes down to, is there some benefit a user would see if an app was trying to serve an ad over HTTP and the app received connection refused vs SSL protocol error or any SSL error.  For me,  the answer is no.  I can't see an app doing something different for different errors in loading an AD.  Either they can load it (or think they can in the case of 80 webserver) or they can't.  
Bottom line, i see no value in it.  I'd propose that if I add the ipv6 option, then at that time the 443 be removed.  This is your call though I'm just trying to do the legwork in providing raw information so you can make a decision.    I'll try to see if i can find at least one real world example of an app using ads over 443, they aren't common I'd expect for the overhead.

**Comment by 0-kaladin** [comment]
And not that my last reply wasn't long enough but regarding IPv6.  I completely agree at this stage it should maybe just be a pref.  I've read articles where ad networks fully support ipv6 they really just aren't utilizing yet.  But if/when they start, the current adaway can do nothing to stop it.  And since as you pointed out there isn't even a hosts file specific for ipv6 ads, they literally just duplicated everything then I can easily enough do that when writing the hosts file.  Although it doubles the hosts file size, it doesn't double the parse side so will still move quickly.

**Comment by dschuermann** [comment]
Wow, thanks for these investigations :)

oom: Thanks for re-adding, could have also happened to me if I would have reworked it, so no worries, I just remembered that we had it implemented.

SSL: I agree to removing 443, looks like a well thought out decision.

IPv6: Yup, I think a pref is great.

**Comment by 0-kaladin** [comment]
Here is the preference commit: https://github.com/Free-Software-for-Android/AdAway/commit/924fdd79f491b6f20fd42728a78eea83feec0b9d

If you have a moment I'd appreciate a code review on that.

**Comment by dschuermann** [comment]
great looks good!

**Comment by 0-kaladin** [comment]
Just fyi, I've posted Adaway in here: www.xda-developers.com/xda-labs/
A ton of people are in xda thread and this gives them another good way to
get the app besides just F-Droid.

**Comment by dschuermann** [comment]
@0-kaladin nice
